### PR TITLE
Un-xfail tests which needed fancy indexing; add `.view`

### PR DESCRIPTION
**Pull request by ev-br**
Add  a `.view` method so that we can remove several xfails.

### CODE REVIEW

**Comment by ev-br** [comment]
On CI this fails in  `__setitem__`, in the ndarrays_to_tensors logic, when doing `tensor[slice] = list`:

```
self = <test_numeric.TestBoolCmp object at 0x7f5910f1f460>

    def setup_method(self):
        self.f = np.ones(256, dtype=np.float32)
        self.ef = np.ones(self.f.size, dtype=bool)
        self.d = np.ones(128, dtype=np.float64)
        self.ed = np.ones(self.d.size, dtype=bool)
        # generate values for all permutation of 256bit simd vectors
        s = 0
        for i in range(32):
>           self.f[s:s+8] = [i & 2**x for x in range(8)]

torch_np/tests/numpy_tests/core/test_numeric.py:464: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = array_w([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1...., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1.])
index = slice(0, 8, None), value = (0, 0, 0, 0, 0, 0, ...)

    def __setitem__(self, index, value):
        index = _helpers.ndarrays_to_tensors(index)
        index = ndarray._upcast_int_indices(index)
        value = _helpers.ndarrays_to_tensors(value)
>       return self._tensor.__setitem__(index, value)
E       TypeError: can't assign a tuple to a torch.FloatTensor
```

Any suggestions @honno?

**Comment by honno** [comment]
To me it looks like something we need to wrap with an additional step, as its saying float tensors can't handle an input NumPy can

```python
In [1]: t = torch.zeros(256)
[PYFLYBY] import torch

In [2]: t[0:8] = [0. for _ in range(8)]
TypeError: can't assign a list to a torch.FloatTensor

In [3]: x = np.zeros(256)
[PYFLYBY] import numpy as np

In [4]: x[0:8] = [0. for _ in range(8)]
```

**Review by lezcano — APPROVED** [review]

**Comment by honno** [comment]
> so maybe instead `asarray()` selectively on flat sequence objects...

Something like

```py
from typing import Sequence
...
   def __setitem__(...):
        ...
        if isinstance(value, Sequence) and all(
            isinstance(e, (int, float, complex)) for e in value
        ):
            value = asarray(value).get()
        ...
```

**Comment by lezcano** [comment]
How does NumPy process these nested sequences into an array? It may be best to simply copy their algorithm.

**Comment by ev-br** [comment]
What numpy actually does is well hidden in its C code. Maybe this test is of help: https://github.com/numpy/numpy/blob/main/numpy/core/tests/test_indexing.py#L808

File: torch_np/_ndarray.py [hunk 1]
```diff
@@ -155,6 +155,13 @@ def copy(self, order="C"):
         tensor = self._tensor.clone()
         return ndarray._from_tensor_and_base(tensor, None)
 
+    def view(self, dtype):
```
[the remark this answers is not in the document]
Comment (ev-br) [reply torch_np/_ndarray.py:158]: Yes. I was holding this refactor to minimize the number of conflicts for gh-23. Now that it's in, it's about time indeed.

**Comment by lezcano** [comment]
My point is, this logic must be somewhere, in Python, C, or whichever language. We may be able to carve it out and implement it here in Python. That's what we did for plenty of functions in PrimTorch really.
Sometimes this is more difficult than abstracting out the ideas and implementing those, but what we found was that this is often more precise than trying to abstract things out.

**Comment by ev-br** [comment]
OK, this PR has run its course. `ndarray.view` is in main, fancy indexing is TODO, best tackled in one go not piecemeal.

**Pull request closed without merging**